thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+        Set<Container> rootSet = getRootSet(id_table);
+
+        pruneEmptyContainers(rootSet);
+
+        Set<Container> finalSet = groupBySubject(rootSet);
+
+        assignThreadIDs(finalSet, threadIDPrefix);
+    }
+
+    /**
+     * Walks the list of emailMessages creating a Container object for each
+     * unique message ID found. Adds the emailMessage to the container where
+     * possible.
+     *
+     * @param emailMessages
+     *
+     * @return - HashMap of all message where the key is the message-ID of the message
+     */
+    private HashMap<String, Container> createIDTable(List<EmailMessage> emailMessages) {
+        HashMap<String, Container> id_table = new HashMap<>();
+
+        for (EmailMessage message : emailMessages) {
+            String messageID = message.getMessageID();
+
+            // Check the id_table for an existing Container for message-id
+            Container container = id_table.get(messageID);
+
+            // An existing container for message-id was found
+            if (container != null) {
+                // If the existing Container has a message already assocated with it
+                // emailMessage maybe a duplicate, so we don't lose the existance of 
+                // the duplicate message assign it a bogus message-id
+                if (container.hasMessage()) {
+                    messageID = String.format("<Bogus-id: %d >", bogus_id_count++);
+                    container = null;
+                } else {
+                    container.setMessage(message);
+                }
+            }
+
+            if (container == null) {
+                container = new Container(message);
+                id_table.put(messageID, container);
+            }
+
+            processMessageReferences(message, container, id_table);
+        }
+
+        return id_table;
+    }
+
+    /**
+     * Loops throught message's list of references, creating objects as needed
+     * and setting up the parent child relationships amoung the messages.
+     *
+     * @param message   The current email messags
+     * @param container Container object for message
+     * @param id_table  Hashtable of known message-id\container pairs
+     */
+    void processMessageReferences(EmailMessage message, Container container, Map<String, Container> id_table) {
+        List<String> referenceList = message.getReferences();
+
+        // Make sure the inReplyToID is in the list of references
+        String inReplyToID = message.getInReplyToID();
+        if (inReplyToID != null && !inReplyToID.isEmpty()) {
+            if (referenceList == null) {
+                referenceList = new ArrayList<>();
+            }
+
+            referenceList.add(inReplyToID);
+        }
+
+        // No references, nothing to do
+        if (referenceList == null) {
+            return;
+        }
+
+        Container parent_ref = null;
+        Container ref;
+
```

Comment:
We should add a comment that referenceList is sorted by parent order.

Reply:
See above comment.